src/trix/models/composition.coffee

```diff
@@ -78,16 +78,22 @@ class Trix.Composition extends Trix.BasicObject
     block = document.getBlockAtIndex(index)
 
     if block.getBlockBreakPosition() is offset
-      document = document.removeTextAtRange(range)
-      range = [position, position]
+      if block.text.getStringAtRange([offset - 1, offset]) is "\n"
+        document = document.removeTextAtRange([position - 1, position])
+      else if offset - 1 isnt 0
+        position += 1
     else
       if block.text.getStringAtRange([offset, offset + 1]) is "\n"
         range = [position - 1, position + 1]
       else if offset - 1 isnt 0
         position += 1
 
     newDocument = new Trix.Document [block.removeLastAttribute().copyWithoutText()]
-    @setDocument(document.insertDocumentAtRange(newDocument, range))
+    if "header" in block.attributes
```

Comment:
We generally don't access properties like `block.attributes` directly and favor the reader methods instead: `block.getAttributes()`.